feat(common): add automatic srcset feature to ngOptimizedImage (#47547)
angular/angular: aio/content/guide/image-directive.md

> ### Handling `srcset` attributes

Reviewer: ```suggestion
### Requesting images at the right size
```

> 'NgOptimizedImage' generates the [`srcset` attribute](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/srcset) for you automatically, based on the configured image loader. The generated `srcset` is dependent on the [`sizes` attribute](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/sizes) passed to the directive. If `sizes` is omitted, it is assumed that your image is meant to be a fixed size, and a srcset is generated with density descriptors (e.g. `1x`, `2x`). If you include a `sizes` property, then a responsive srcset is provided, with a wide array of possible sizes. By default, these sizes are:

Reviewer: Can we add a few lines to explain what srcset/sizes mean? As I was trying to provide some suggestions below, I'm wondering if we're asking for folks to learn too much about how "sizes" works (and it's kind of complicated!). If we had a higher-level API (like `responsive="100vw"`), it would be easier to learn and we could fall through to "sizes".

Anyway, here's a stab at adding more info about srcset/sizes, but I think we'd really need to add more context:

```suggestion
Defining a [`srcset` attribute](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/srcset) ensures that the browser requests an image at the right size for your user's viewport, so it doesn't waste time downloading an image that's too large. 'NgOptimizedImage' generates an appropriate `srcset` for the image, based on the presence and value of the [`sizes` attribute](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/sizes) on the image tag.

#### Fixed size images

If your image should be "fixed" in size  (i.e. the same size across devices), there is no need to set a `sizes` attribute. A `srcset` can be generated automatically from the image's width and height attributes with no further input required. 

Example srcset generated: `<img ... srcset="image-400w.jpg 1x, image-800w.jpg 2x">`

#### Responsive images

If your image should be responsive (i.e. grow and shrink according to viewport size), then you will need to define a [`sizes` attribute](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/sizes) to generate the `srcset`.

If you haven't used `sizes` before, a good place to start is to set it based on viewport width. For example, if your CSS causes the image to fill 100% of viewport width, set `sizes` to `100vw` and the browser will select the image in the `srcset` that is closest to 100% vw (after accounting for pixel density). If your image is only likely to take up half the screen (ex: in a sidebar), set `sizes` to `50vw` to ensure the browser selects a smaller image. And so on.

By default, these sizes are:
```

Viewport sizes: `[640, 750, 828, 1080, 1200, 1920, 2048, 3840]`
Sub-Viewport sizes (only used if the sizes property is less than 100vw): `[16, 32, 48, 64, 96, 128, 256, 384]`

> If you would like to customize either or both of these breakpoint arrays, you can do so using the `IMAGE-CONFIG` provider:

Author: Fixed

<code-example format="typescript" language="typescript">
providers: [
  {
    provide: IMAGE_CONFIG,
    useValue: {
      viewportBreakpoints: [640, 750, 828, 1080, 1200, 1920],
      subViewportBreakpoints: [16, 48, 96, 128, 384]
    }
  },
],
</code-example>

If you would like to manually define a`srcset` attribute, use the `ngSrcset` attribute, and automatic `srcset` generation will be disabled:

<code-example format="html" language="html">

&lt;img ngSrc="hero.jpg" ngSrcset="100w, 200w, 300w"&gt;

</code-example>

> If the `ngSrcset` attribute is present, `NgOptimizedImage` generates and sets the  using the configured image loader. Do not include image file names in `ngSrcset` - the directive infers this information from `ngSrc`. The directive supports both width descriptors (e.g. `100w`) and density descriptors (e.g. `1x`) are supported.

Reviewer: How do `ngSrcset` and the generated `srcset` intersect? Is there any reason to keep `ngSrcset` around if you can define custom breakpoints anyway?

Author: Right now if you use `ngSrcset` the directive will just use that value, and won't use the automatic srcset. I'm not sure if we need to keep providing the ngSrcset convenience API or not. It could be useful for some people, but it also does complicate the API

<code-example format="html" language="html">

&lt;img ngSrc="hero.jpg" ngSrcset="100w, 200w, 300w" sizes="50vw"&gt;

</code-example>

### Disabling automatic srcset generation

To disable srcset generation for a single image, you can add the `unoptimized` attribute on the image:

<code-example format="html" language="html">

&lt;img ngSrc="about.jpg" unoptimized&gt;

</code-example>

>  Alternatively, you can set the `disableAutoSrcset` property to `true` on the IMAGE_CONFIG provider:
> 
> <code-example format="typescript" language="typescript">
> providers: [
>   {
>     provide: IMAGE_CONFIG,
>     useValue: {
>       disableAutoSrcset: true
>     }
>   },
> ],
> </code-example>
> 

Reviewer: We agreed to remove this feature, no?

```suggestion

```
